### DISCUSSION TITLE
Reopening closed issues

### DISCUSSION THREAD

**Discussion by CyrilleB79** [nvaccess/nvda#15551] [General]
Hello all

I have had a recent discussion with @britechguy , first publicly in comments of #13021 and #14541 and then through private e-mail exchanges. The discussion is about reopening closed issues.

There are issues that have been closed, for various reasons. But sometimes, the problem encountered by users and that had triggered them to report the issue comes back again and again though. This probably means that the people looking at issues (developers, triagers) have not well understood the root problem or have not been able to explain to the users/reporter why the problem culd not be solved by a modification in NVDA's code or documentation.

I have encouraged   @britechguy  to open a new issue with only a summary of the information that is still useful, up-to-date and relevant with a link to the old closed issue. From my experience, it seemed to me the most efficient to progress. But from a user/reporter point of view, @@britechguy disagree with this way to do; @britechguy, I let you explain your point of view yourself, you will do it better than I.

I realize that all the contributing documentation of NVDA does not seem to provide instructions when an issue is closed but is not resolved from a user's point of view. Or do I miss something?
We must remain attentive to making users feel heard.

NV Access' opinion (@seanbudd, @michaelDCurran) would be appreciated, as well as active contributors' ones (@leonardder, @josephsl, @XLTechie, @lukaszgo1, @Adriani90).
Also active reporters opinion is needed (@britechguy, @cary-rowen).

I have no other name in mind but feel free to mention other people.

After this discussion has taken place, I hope that the contributing documentation will be updated accordingly.

Thanks.
Cheers,

Cyrille

**Comment by britechguy** [comment]
I am simply going to include pertinent segments of what I said to Cyrille privately:
----
My days of opening new issues, that are essentially direct extensions of existing, closed issues, are over.  It's too damned much work and it's too inefficient for those on "the commenting" or "the opening" end.  When the same thing comes up, again and again, and there exist issues for them, my opinion is that those should be reviewed for reopening if/when someone chooses to comment on them.  They do that because what they've read is relevant to what they're about to say.  It's also not, in my opinion, at all efficient to have issues that make reference to other issues when that can be avoided.

If someone's trying "to raise an issue from the dead" it's because they still have it.
----
If the semi-official policy is to brush aside those who attempt to revive a closed topic, I stand by my assessment that this is a mistake.  I have been trying, for years now, to get the users of NVDA who post to the NVDA Group to get on to GitHub and communicate with NVDA developers.  I routinely tell them to search for existing issues, and comment on same if they exist as opposed to opening new ones.

Activity on an issue, regardless of its status as open or closed, indicates it needs attention.  It's no different than entering a new issue in that it's an attempt to get attention to some need.  If those who are in the position to triage issues (and I don't just mean you) don't look at what the comment states, and how it ties in to what's already part of a closed issue, as part of considering whether it should be reopened or not, I really don't know what to say.  

I'm capable, very capable, of creating new issues, yet even I don't like doing that when I'm just rattling the same proverbial cage that someone else has rattled in the past.  If I comment on a closed issue, it's because what's in that issue was never resolved and it's directly linked to what I feel needs attention because I'm having the same issue.
----


When it comes right down to it, it's inefficient for all involved to have new issues alluding to one or more old issues when those happen to be about what appears to be exactly the same thing.  It's the equivalent of topic-splitting in forums, where the same topic is being discussed, often in parallel, because someone decided to create another one without having checked whether there is a current one out there.  The difference being that when it comes to forums, reviving many things from the dead does make no sense.  But in a project management system, when a search on the issue you intend to report indicates it's been reported before, it makes far more sense to raise the flag on those prior reports, whether the issue was closed or not.

I see little difference in issues that have been allowed to remain open, sometimes for years (and I was just commenting recently along with a number of others, and it was opened in 2016), versus one that was closed, often with no clear rationale as part of the closure.

And note well, I'm not talking issues closed because they are fixed, or where they are closed because they are duplicates of another issue and where the closure message makes direct reference to where any further follow-up is required.

The time of reporters is limited, too, and "reinventing the wheel" rather than saying, "Me, too, now - and here are my details," on an existing issue is often going to discourage actual communication from the user community.  One of NVDA's great strengths is that a mechanism exists for end users to have direct communication with the development team.  Making that as easy as possible needs to be a significant consideration.

**Comment by lukaszgo1** [comment]
I'd say there is no single good answer here, an a lot depends on why the given issue was closed. Most of the time this is because:
- someone triaging misunderstood the problem (just an user error, or, way too often IMHO, lack of technical competence on part of the person doing triage). In that case the issue should be simply reopened no question asked.
- The problem was not fully solved - in that case new issue should be created listing exactly what remains to be done. This is by no means making it harder for reporters. Time of developers able to contribute a fix is limited too, and if we can avoid reading (often quite long) trail of comments we are way more likely to contribute a fix.
- The last  category (both issues linked above fall there) is simply the case in which someone disagrees with the reason for closing. I don't think we can come up with any hard rules there, and these should be considered on a case by case basis.

For the issues where core developers do not intent to accept a change of behavior we can consider simply locking the conversation, though that can be considered too strict by some.

**Comment by britechguy** [comment]
@lukaszgo1 

I agree entirely that there is no single good answer here.

For issues where core developers do not intend to accept a change in behavior, the closure note should include that.  Most of us who've "been around the block" can accept a, "No, we're not doing that," or a, "No, we're not doing that with detailed rationale included."  What is very hard to swallow is a No with either a weak rationale that does not address the issue, as written, or just a closure without any comment.

I know better than many how little spare time developers have, and I respect that, but when an issue is closed some reason, even if it's a straight, "We do not intend to implement this now, or in the future," should be supplied.

As to the "trail of comments," I personally, when I'm the developer, prefer to have them there for referencing if I need or want to.  I'm perfectly capable of zooming straight to what's just been added, first, then deciding if I need or want any of the back story.  But when I do, I find it much easier to have it there, at the ready, than having to go through multiple levels of redirection to other issues, and the like.  That's part of why I said that if a user lands upon a previously closed issue, and it really is a direct reflection of what they believe their current issue to be, that a comment on that should trigger a review, and a sincere one.  It feels like reinventing the wheel to report, in its entirety, something that's already been reported and where you need only add something as simple as a "me, too, now" and/or how what's going on for you may differ, slightly, from what's come before.  A variation on the central theme, as it were.

**Comment by Brian1Gaff** [comment]
The only thing I'd say is that sometimes we do not understand the ramifications of what we ask for. I do agree that a simple non technical non patronising explanation can go a long way to preserve trust that the decision was not made lightly, ignoring an issue.
 Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.

**Comment by britechguy** [comment]
> a simple non technical non patronising explanation can go a long way to preserve trust that the decision was not made lightly, ignoring an issue.

And that is the crux of the matter.  If someone has taken the time to create an issue, those closing them should at least try to give some rationale for the closing if it's not being closed because the requested fix has been made.

End users (and I'll include myself, too) often make requests that are not feasible to implement for one or more reasons.  Having a one or two liner saying something along these lines, and being specific but not overwhelmingly technical when possible, goes a long way toward preserving trust and not turning people off to reporting what they believe to be other issues in the future.  I have been saying, for years, that NVDA's end users are blessed because it is one of the few instances where an established direct communication line exists for end users to developers.

Giving a rationale also goes a long way toward preventing someone from resurrecting certain issues from the dead if the explanation for why the request is not feasible makes sense to them.  It'll sometimes stop them right in their tracks and prevent the creation of another issue or feature request that would have the same fate.  This cuts down on work for all who are fielding these requests.

**Comment by CyrilleB79** [comment]
I agree that if issues are closed without a fix going to NVDA, a justification should be provided. And as Brian says, the justifications should be understandable at least by the person who has opened the issue. I think that it's done for most of the issues, and specifically for #13021 and #14541 that I have cited as example.

However there are issues (maybe 1 out of 4) for which such justification is not provided. In this case, users/reporters are more than legitimate to ask for a clarification in the closed issue; and as an issue reporter, I have already done so and received an answer. But many users/reporters will probably not dare asking for this clarification. That's why the triager is primarily responsible for providing this clarification when closing.


Also some reactions to what was written before, especially by Brian:
> I routinely tell them to search for existing issues, and comment on same if they exist as opposed to opening new ones.

That's a good point if open issues are found. That means that no duplicate will be opened and that the discussion will take place at the same location. Regarding closed issue, there is of course no good and unique solution. Especially when the reporter disagree with the reason why it was closed, we should take care that the issue does not become a forum thread.

> Activity on an issue, regardless of its status as open or closed, indicates it needs attention. 

If users try to interact with developers/triagers, it should of course require attention. But users reporting issues should be aware that an open issue means some work to do for developers; and a closed issue means no more work to do on this task. These issues are first a way to organize the work. On the opposite, forum threads are primarily discussions and thus, remain open forever.
To illustrate with a similar situation: in my company, the IT manages the issues reported by users through an issue tracking system. Once the issue is considered resolved by the IT, the issue ticket is close. And as an user, I have one week to accept or decline the solution, i.e. I have one week to re-open the issue.
Of course, for NVDA we probably do not need to be so strict but this illustrates that re-opening discussions after many months on a closed issue is not desirable for developers.

So maybe instructions/checks should be provided to users who would like to re-open closed issues:
* Have you understood why the issue was closed? (if not, ask for clarification)
* Do you disagree with the reason why it was closed? (If yes, please provide clear and strong arguments explaining why the reason is not relevant)
* etc.

**Reply by Chealer** [reply]
Greetings Cyrille,

> If users try to interact with developers/triagers, it should of course require attention. But users reporting issues should be aware that an open issue means some work to do for developers; and a closed issue means no more work to do on this task.

A couple cosmetic remarks:
1. Issues themselves are not “open” or “closed”. (I can see that GitHub writes “Close issue”, but it should not be a reference with regards to terminology.)
2. An issue _is_ not a task. It very often maps to one, but resolving an issue could require multiple tasks. And some issues would be so prohibitive to solve that there's (arguably) no associated task.

That being said, thank you for bringing up this [meta-]issue. I can guess the kind of meta-issues @britechguy is dealing with.

But it is not correct that a ticket marked as closed means the issue it tracks no longer requires work. As far as I know, GitHub’s Open/Closed labels are comparable to what it calls labels and have no specific/intrinsic meaning. I am not even convinced that issues with tickets marked as Closed need less work on average than those with tickets marked as Open.

I think this comes down to expectations. NVDA is FLOSS which is not sold AFAIK. The ITS you are struggling with is powered by GitHub, which is SaaS offered for free. Its main strength is its VCS, while its ITS engine “KISSes”. Its computational performance is good, but the performance it allows to its users is another story. It is more targeted at amateur projects than major projects of NVDA’s size. With thousands of issues, it becomes a mess. In ticket #19969, I reported one of its worst issues, which visibly contributes to the difficulties discussed here, but I do not expect the project will solve that before months of considerable investment.😐

> To illustrate with a similar situation: in my company, the IT manages the issues reported by users through an issue tracking system. Once the issue is considered resolved by the IT, the issue ticket is close. And as an user, I have one week to accept or decline the solution, i.e. I have one week to re-open the issue.
> Of course, for NVDA we probably do not need to be so strict but this illustrates that re-opening discussions after many months on a closed issue is not desirable for developers.

I don't work for your company, but I remember similar behavior in the last ministry I served. The policy said once your issue was marked as resolved, you had 3 days to infirm. If that happened while you were on vacation or on leave, too bad― just file a new ticket, and go explain again to another―hopefully, but rarely, more competent―technician.🙄

The reason for such policies is indicatorism. Nowadays, even governments mandate the use of [performance dashboards intended to show that things are going even better](https://performance.gouv.qc.ca/accueil/). Managers subjected to this kind of evaluation often feel pressured to game the system to corrupt the indicators in their favor. It looks better to have solved “3 issues” in an average of 2 weeks than to have solved 1 issue in 3 months.

I saw this in the public sector of a parliamentary democracy, but I am sure public companies are not immune to [indicatorism](https://en.wikipedia.org/wiki/Goodhart%27s_law).

As a developer or technician, restarting from 0 is the last thing you want. You want as much information about what happened, what the customer said, what your team tried, when the problem started, etc. (Unless of course, your own performance is evaluated based on such metrics. In that case, you will be happy to “resolve” the same issue as often as needed.😁)

> So maybe instructions/checks should be provided to users who would like to re-open closed issues:
> 
> * Have you understood why the issue was closed? (if not, ask for clarification)
> 
> * Do you disagree with the reason why it was closed? (If yes, please provide clear and strong arguments explaining why the reason is not relevant)
> * etc.

I see Open and Closed labels as noise, pretty much like what GitHub calls labels. They should be ignored rather than studied. Except if you think a ticket being "closed" has a negative impact, then _that_ is a meta-issue which should be discussed/reported.

[🅭🄍](https://www.philippecloutier.com/Common+infrastructure+licensing#forum)

**Reply by britechguy** [reply]
@Chealer wrote, in part:  Issues themselves are not “open” or “closed”.

I beg to differ.  When you go to the NVDA issues page, the search box is pre-populated with [is:issue state:open].  Only issues in open status (which is different than being actively worked status, I'll grant you) are shown.

You can also change the state for an issue search to "closed" instead of "open."  I know of nowhere in the software project management world where "closed" means anything other than, "Not being worked, and will not be worked in the future."

If someone reports something sufficient to reopen an issue, and I'll use an very recent example of what might be just such a thing that @CyrilleB79 did:  https://github.com/nvaccess/nvda/issues/19952.  That issue has not, as yet, been reopened.  And if it were to remain closed, the only reasonable expectation is that the information offered post-closure is not considered enough to trigger anything else to be examined or worked on.  A closed issue is just that: closed and done.

I have no argument with the idea that a single issue may involve many different tasks.  But a closed issue should have no open tasks associated with it.  Closed, completed, done, or "not planned" all mean, effectively, the same thing:  nothing actionable now or in the foreseeable future.

**Reply by Chealer** [reply]
Greetings Brian,

> @Chealer wrote, in part: Issues themselves are not “open” or “closed”.
> 
> I beg to differ. When you go to the NVDA issues page, the search box is pre-populated with [is:issue state:open]. Only issues in open status (which is different than being actively worked status, I'll grant you) are shown.

I realize my comment could have been clearer. That remark was purely terminological. I simply meant that you don't “open an issue”; you open/file a _ticket_ which tracks an issue, or you create/open a _request_.

> You can also change the state for an issue search to "closed" instead of "open." I know of nowhere in the software project management world where "closed" means anything other than, "Not being worked, and will not be worked in the future."

There are many many software projects, each with their own terminology (jargon😏?), more or less aligned with other projects. There may well be projects where that is correct, but there are many which use it differently. Just looking at NVDA issues reported this month, and focusing on the minority which I can grasp, I see quite a few cases which don't match that definition:
1. Issue #19919 
2. Issue #19930
3. Issue #19947 

The meaning you propose would have little value, since we don't know if an issue will be worked on in the future.

> If someone reports something sufficient to reopen an issue, and I'll use an very recent example of what might be just such a thing that @CyrilleB79 did: #19952. That issue has not, as yet, been reopened. And if it were to remain closed, the only reasonable expectation is that the information offered post-closure is not considered enough to trigger anything else to be examined or worked on. A closed issue is just that: closed and done.

To me, these qualifiers don't apply to issues. A task can be done, but not an issue.

> Closed, completed, done, or "not planned" all mean, effectively, the same thing: nothing actionable now or in the foreseeable future.

My next trip is not planned, but it's far from completed. As for "closed", that depends on which project you ask. I would even say that depends on who you ask… different contributors may have different usages.

Basically, a proper ITS tracks the project's workflow, which can involve a lot of statuses depending on the project's complexity. GitHub is free, but it only offers a boolean. It won't do wonders.

You do have a very good point about the search box’s default value. I've discovered GitHub’s ITS engine so many years ago that I don't think about it anymore and can't remember exactly how it was initially, but the behavior at least in this project at this time is doubtful. “Open” and “Closed” are top left, but besides that, they don't look like tabs at all. I also notice that the number 2169 shows next to Open, and 12255 after Closed. But if I'm in the Code tab, the “number” next to Issues is “2.2k”.

I would say issues which only have closed tickets are less discoverable than those which have open tickets, and it looks like the ITS is not considering closed tickets on par with open tickets.

**Comment by CyrilleB79** [comment]
@Chealer, please keep focused on NVDA specific project. If you wish to criticize more broadly GitHub (process, terminology) do it directly at GitHub level, not in NVDA project. Thanks.

Also this thread is about how to manage closed issues (using the usual terminology of GitHub and NVDA project). Let's remain focused on this topic.

**Reply by Chealer** [reply]
Greetings Cyrille,

> @Chealer, please keep focused on NVDA specific project. If you wish to criticize more broadly GitHub (process, terminology) do it directly at GitHub level, not in NVDA project.

The point should not be to criticize GitHub or anything. The point is to find solutions to the problem(s) you are reporting.

> Also this thread is about how to manage closed issues (using the usual terminology of GitHub and NVDA project).

See [my earlier comment](https://github.com/nvaccess/nvda/discussions/15551#discussioncomment-16606423)

> Let's remain focused on this topic.

_Slowly, but surely_. There is much literature (and even methodologies) which emphasize the importance of starting with the problem. Notably, in [DMAIC](https://en.wikipedia.org/wiki/DMAIC), identifying solutions only comes in _Improve_ (step #4).

> A problem well defined is a problem half solved.

**Reply by CyrilleB79** [reply]
> > Also this thread is about how to manage closed issues (using the usual terminology of GitHub and NVDA project).
> 
> See [my earlier comment](https://github.com/nvaccess/nvda/discussions/15551#discussioncomment-16606423)

Yes, I've seen it.
And I disagree with the need to question the meaning of terms that are commonly used and understood in NVDA GitHub community as well as in GitHub itself. Do not expect me to discuss it further: to me this specific terminology discussion is useless, and in any case, at the wrong place.

**Reply by Chealer** [reply]
Greetings Cyrille,

> Yes, I've seen it. And I disagree with the need to question the meaning of terms that are commonly used and understood in NVDA GitHub community as well as in GitHub itself.

There are many terms which are commonly used and "understood" in many communities, but which are nevertheless commonly problematic. The problem is most terms have multiple meanings, so one person's understanding may differ from another's. The term _[hood](https://en.wiktionary.org/wiki/hood)_, for example, has ~10 to 20 senses, just as a common noun. So, when one of my colleagues and his former colleague were told to remove a vehicle's hood, they removed the _capot_. Unfortunately, to the UK officer who had ordered that, a _capot_ was a _bonnet_, and the _hood_ was the vehicle's _roof_.😬

If you're going to have a policy debate about hood removal, what will help most will not be using the most commonly used terms, but rather those which are reliably understood _in the intended sense_. Unfortunately, the word "open" is _way_ more polysemous/vague than the already problematic "hood", so reaching any consensus will remain elusive without a clear definition in context.

> Do not expect me to discuss it further: to me this specific terminology discussion is useless, and in any case, at the wrong place.

I'm afraid you're confusing discussions here. Brian may have launched a terminological discussion, but the topic of my (first) message was **expectations**, not terminology. _Free_ as in beer does not mean _complication_-free. Gratis does not mean no total cost of ownership.

Speaking about confusion, while nobody expects you to discuss this further, if you choose to reply, please make it clear which message each of yours replies to.

**Reply by CyrilleB79** [reply]
> I'm afraid you're confusing discussions here.

No.

> the topic of my (first) message was **expectations**, not terminology. 

Discussion around terminology in your first message here, as well as in other messages of yours in here and in this repo, takes an important place, actually a too important place in my opinion. That's why I do not plan to question terminology as you do.

> _Free_ as in beer does not mean _complication_-free. Gratis does not mean no total cost of ownership.

What do you mean? NVDA is free software but managed by an organisation, NV Access, of which I am not a member by the way.

> Speaking about confusion, while nobody expects you to discuss this further, if you choose to reply, please make it clear which message each of yours replies to.

If any of my replies are unclear in terms of which message they refer to, please point to the specific one so I can clarify.

**Reply by Chealer** [reply]
> > I'm afraid you're confusing discussions here.
> 
> No.

Well, you may think you are not confusing discussions, but I _do_ remain unconvinced that you are.

> > the topic of my (first) message was **expectations**, not terminology.
> 
> Discussion around terminology in your first message here, as well as in other messages of yours in here and in this repo, takes an important place, actually a too important place in my opinion. That's why I do not plan to question terminology as you do.

Heh, you are free to "not question terminology”, as you put it, but then don't be surprised to get "cosmetic remarks" if you choose to use confusing terms, in particular when you do that in a discussion about tracking hygiene/processes which you started. In any case, if you think there is anything wrong with the share of a message dedicated to terminology or any other topic, the constructive thing to do would be to comment on that specific message.

> > _Free_ as in beer does not mean _complication_-free. Gratis does not mean no total cost of ownership.
> 
> What do you mean? NVDA is free software but managed by an organisation, NV Access, of which I am not a member by the way.

Right, NVDA is offered for free by a non-profit organization. That means its infrastructure budget is limited, and it chose free SaaS to power its ITS. One of the worst consequences of that tradeoff is workflow/resolution tracking. There is only so much you can expect from an ITS engine which has little more than a boolean to do all that.

I must clarify I am not saying that the current situation is entirely acceptable. The first example you provide, ticket #13021, was clearly mismanaged (marked as completed in 2021, when the job was only actually completed >4 years later), which understandably creates frustration and confusion.

I think an efficient exercise to get to the root of what you noticed would be to provide a few cases/tickets where you feel something is wrong but there is no consensus on the best course.

> > Speaking about confusion, while nobody expects you to discuss this further, if you choose to reply, please make it clear which message each of yours replies to.
> 
> If any of my replies are unclear in terms of which message they refer to, please point to the specific one so I can clarify.

I was just offering general advice. The only problematic message I saw was your April message this one replies to.

**Reply by seanbudd** [reply]
Please stop wasting our community's time with off-topic discussion.
Please respect community member requests to stop haranguing them i.e. respect Cyrille's previous message.

>Do not expect me to discuss it further: to me this specific terminology discussion is useless, and in any case, at the wrong place.
